src/main/kotlin/org/jmailen/gradle/kotlinter/KotlinterPlugin.kt

```diff
@@ -20,146 +20,121 @@ import org.jmailen.gradle.kotlinter.tasks.LintTask
 class KotlinterPlugin : Plugin<Project> {
 
     val extendablePlugins = mapOf(
-            "org.jetbrains.kotlin.jvm" to KotlinJvmSourceSetResolver,
-            "kotlin-android" to AndroidSourceSetResolver)
-
-    override fun apply(project: Project) {
-        val kotlinterExtension = project.extensions.create("kotlinter", KotlinterExtension::class.java)
-
-        // for known kotlin plugins, create tasks by convention.
-        val taskCreator = TaskCreator(project)
-        extendablePlugins.forEach { pluginId, sourceResolver ->
-            project.plugins.withId(pluginId) {
-
-                sourceResolver.applyToAll(project) { id, files ->
-                    taskCreator.createSourceSetTasks(id, files)
-                }
-
-                taskCreator.createParentTasks()
-            }
-        }
-
-        project.afterEvaluate {
-            val ktLintParams = KtLintParams(
-                kotlinterExtension.indentSize,
-                kotlinterExtension.continuationIndentSize,
-                kotlinterExtension.experimentalRules,
-                kotlinterExtension.disabledRules,
-                project.editorConfigPath()
-            )
-
-            taskCreator.lintTasks.forEach { lintTask ->
-                lintTask.ignoreFailures = kotlinterExtension.ignoreFailures
-                lintTask.reports = kotlinterExtension.reporters.associate { reporter ->
-                    reporter to project.reportFile("${lintTask.sourceSetId}-lint.${reporterFileExtension(reporter)}")
+        "org.jetbrains.kotlin.jvm" to KotlinJvmSourceSetResolver,
+        "kotlin-android" to AndroidSourceSetResolver
+    )
+
+    override fun apply(project: Project) = with(project) {
+        val kotlinterExtension = extensions.create("kotlinter", KotlinterExtension::class.java)
+
+        // for known kotlin plugins, register tasks by convention.
+        extendablePlugins.forEach { (pluginId, sourceResolver) ->
+            pluginManager.withPlugin(pluginId) {
+
+                val lintKotlin = registerParentLintTask()
+                val formatKotlin = registerParentFormatTask()
+
+                sourceResolver.applyToAll(project) { id, resolveSources ->
+                    val lintKotlinPerVariant = tasks.register("lintKotlin${id.capitalize()}", LintTask::class.java) { lintTask ->
```

Comment:
nit: `lintTaskPerSourceSet` would be more accurate -- variant is android specific terminology I believe.